### PR TITLE
changes around solution crawler.

### PR DESCRIPTION
**Pull request by heejaechang**
first, made solution crawler options to show up in Roslyn/Option pane to make experiement easier.
second, tweaked the way higher/normal/low priority processor interact each other.
last, made solution crawler to pick next project to process a bit smarter.

### CODE REVIEW

**Comment by heejaechang** [comment]
@srivatsn @shyamnamboodiripad @mavasani can you take a look?

File: src/Features/Core/SolutionCrawler/WorkCoordinator.GlobalOperationAwareIdleProcessor.cs [hunk 1]
```diff
@@ -76,10 +76,10 @@ private void OnGlobalOperationStopped(object sender, GlobalOperationEventArgs e)
                         _globalOperationTask = SpecializedTasks.EmptyTask;
                     }
 
-                    protected async Task GlobalOperationWaitAsync()
+                    protected Task GlobalOperationWaitAsync()
                     {
                         // we wait for global operation if there is anything going on
-                        await _globalOperationTask.ConfigureAwait(false);
+                        return _globalOperationTask;
                     }
```
Comment (tmeschter) [inline src/Features/Core/SolutionCrawler/WorkCoordinator.GlobalOperationAwareIdleProcessor.cs]: Why this change?

File: src/Features/Core/SolutionCrawler/WorkCoordinator.HighPriorityProcessor.cs [hunk 2]
```diff
@@ -114,7 +114,7 @@ protected override Task WaitAsync(CancellationToken cancellationToken)
                             // see whether we have work item for the document
                             WorkItem workItem;
                             CancellationTokenSource documentCancellation;
-                            Contract.ThrowIfFalse(GetNextWorkItem(out workItem, out documentCancellation));
+                            Contract.ThrowIfFalse(TryGetNextWorkItem(out workItem, out documentCancellation));
```
Comment (tmeschter) [inline src/Features/Core/SolutionCrawler/WorkCoordinator.HighPriorityProcessor.cs]: I'm not sure I understand this. Is it OK *in general* for `TryGetNextWorkItem` to return `false`, *in this specific case* we know it won't?

File: src/Features/Core/SolutionCrawler/WorkCoordinator.GlobalOperationAwareIdleProcessor.cs (hunk 1 above)
Comment (heejaechang) [inline src/Features/Core/SolutionCrawler/WorkCoordinator.GlobalOperationAwareIdleProcessor.cs]: because we dont need await.

File: src/Features/Core/SolutionCrawler/WorkCoordinator.HighPriorityProcessor.cs (hunk 2 above)
Comment (heejaechang) [inline src/Features/Core/SolutionCrawler/WorkCoordinator.HighPriorityProcessor.cs]: other case, it has input like documentId and projectId, and we might not have workitem for the given input, but this case, it says any workItem, and there must be one. let me re-name method.

**Comment by heejaechang** [comment]
@srivatsn @mavasani @shyamnamboodiripad can you take a look?

File: src/Features/Core/SolutionCrawler/WorkCoordinator.AsyncDocumentWorkItemQueue.cs [hunk 3]
```diff
@@ -104,6 +95,33 @@ private DocumentId GetBestDocumentId_NoLock(ProjectId preferableProjectId)
...
+                private ProjectId GetBestProjectId_NoLock(ProjectId projectId, ProjectDependencyGraph dependencyGraph)
+                {
+                    if (projectId != null)
+                    {
+                        if (_documentWorkQueue.ContainsKey(projectId))
+                        {
+                            return projectId;
+                        }
+
+                        foreach (var dependingProjectId in dependencyGraph.GetProjectsThatDirectlyDependOnThisProject(projectId))
```
Comment (mavasani) [inline src/Features/Core/SolutionCrawler/WorkCoordinator.AsyncDocumentWorkItemQueue.cs]: Basically we are preferring dependent projects if no pending work for active project, right?

File: src/Features/Core/SolutionCrawler/WorkCoordinator.GlobalOperationAwareIdleProcessor.cs (hunk 1 above)
Comment (mavasani) [inline src/Features/Core/SolutionCrawler/WorkCoordinator.GlobalOperationAwareIdleProcessor.cs]: You should remove the comment above the modified statement and rename the method to GetGlobalOperationTask, or better make it a property GlobalOperationTask so the caller knows its fetching a task.

File: src/Features/Core/SolutionCrawler/WorkCoordinator.LowPriorityProcessor.cs [hunk 4]
```diff
@@ -76,6 +73,39 @@ protected override Task WaitAsync(CancellationToken cancellationToken)
...
+
+                            if (!globalOperationTask.IsCompleted ||
+                                !highPriorityTask.IsCompleted ||
+                                !normalPriorityTask.IsCompleted)
+                            {
+                                await Task.WhenAll(globalOperationTask, highPriorityTask, normalPriorityTask).ConfigureAwait(false);
+                            }
+
+                            // if there are no more work left for higher queue, then it is our time to go ahead
+                            if (!HigherQueueHasWorkItem)
```
Comment (mavasani) [inline src/Features/Core/SolutionCrawler/WorkCoordinator.LowPriorityProcessor.cs]: Do we really need a while loop to do idle processing here, we already have idle processing loop in base type right?

Another approach would be to change this method to return `Task<bool>` and return value `HigherQueueHasWorkItem`. Based on this value, the caller can go ahead with execution if HigherQueueHasWorkItem = false or return early with no-op otherwise?

File: src/Features/Core/SolutionCrawler/WorkCoordinator.NormalPriorityProcessor.cs [hunk 5]
```diff
@@ -183,6 +187,35 @@ public Task Running
                         }
                     }
 
+                    private async Task WaitForHigherPriorityOperationsAsync()
```
Comment (mavasani) [inline src/Features/Core/SolutionCrawler/WorkCoordinator.NormalPriorityProcessor.cs]: Can we refactor some of the common code here to the common base type between low and normal priority processor?

**Comment by mavasani** [comment]
:+1: modulo some minor refactoring suggestions, which you are free to ignore if they don't seem useful/feasible.

File: src/Features/Core/SolutionCrawler/WorkCoordinator.AsyncDocumentWorkItemQueue.cs (hunk 3 above)
Comment (heejaechang) [inline src/Features/Core/SolutionCrawler/WorkCoordinator.AsyncDocumentWorkItemQueue.cs]: yep

File: src/Features/Core/SolutionCrawler/WorkCoordinator.GlobalOperationAwareIdleProcessor.cs (hunk 1 above)
Comment (heejaechang) [inline src/Features/Core/SolutionCrawler/WorkCoordinator.GlobalOperationAwareIdleProcessor.cs]: good idea

File: src/Features/Core/SolutionCrawler/WorkCoordinator.LowPriorityProcessor.cs (hunk 4 above)
Comment (heejaechang) [inline src/Features/Core/SolutionCrawler/WorkCoordinator.LowPriorityProcessor.cs]: we can't return.

File: src/Features/Core/SolutionCrawler/WorkCoordinator.NormalPriorityProcessor.cs (hunk 5 above)
Comment (heejaechang) [inline src/Features/Core/SolutionCrawler/WorkCoordinator.NormalPriorityProcessor.cs]: sure

**Comment by heejaechang** [comment]
@shyamnamboodiripad @srivatsn @JohnHamby can you take a look?

File: src/Features/Core/SolutionCrawler/WorkCoordinator.AsyncDocumentWorkItemQueue.cs (hunk 3 above)
Comment (srivatsn) [inline src/Features/Core/SolutionCrawler/WorkCoordinator.AsyncDocumentWorkItemQueue.cs]: Can you please add comments here about the heuristics?

**Comment by srivatsn** [comment]
:+1:
